entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy

```diff
@@ -197,50 +192,36 @@ declare function {$prefix}:extract-instance-{$entity-type-name}(
     $source as node()?
 ) as map:map
 {{
-    let $source-node := {$prefix}:init-source($source, '{$entity-type-name}')
+    let $source-node := es:init-source($source, '{$entity-type-name}')
+    (: begin customizations here :)
+{ 
+    let $properties := $model
+        =>map:get("definitions")
+        =>map:get($entity-type-name)
+        =>map:get("properties")
+    let $variable-setters := 
+        for $property-name in map:keys($properties)
+        return es-codegen:variable-line-for($prefix, $model, $entity-type-name, $property-name)
+    return fn:string-join( $variable-setters, "&#10;")
+    }
+    (: end customizations :)
 
-    let $instance := json:object()
-    (: If desired, add the original source document as an attachment. :)
-        =>{$prefix}:add-attachments($source-node, $source)
-    (: Add type information to the entity instance (required, do not change) :)
-        =>map:with('$type', '{ $entity-type-name }')
+    let $instance := es:init-instance($source-node, '{ $entity-type-name }')
+    (: Comment or remove the following line to suppress attachments :)
+        =>es:add-attachments($source-node, $source)
 
     return
-
-    (: The following logic may not be required for every extraction       :)
-    (: if this $source-node has no child elements, it has a reference key :)
-    if (empty($source-node/*))
-    then $instance=>map:with('$ref', $source-node/text())
-    (: Otherwise, this source node contains instance data. Populate it. :)
-    else
-
-    (:
-    The following code populates the properties of the '{$entity-type-name}'
-    entity type. Ensure that all of the property paths are correct for your
-    source data.  The general pattern is
-    =>map:with('keyName', casting-function($source-node/path/to/data))
-    but you may also wish to convert values
-    =>map:with('dateKeyName',
-          xdmp:parse-dateTime('[Y0001]-[M01]-[D01]T[h01]:[m01]:[s01].[f1][Z]',
-          $source-node/path/to/data/in/the/source))
-    You can also implement lookup functions,
-    =>map:with('lookupKey',
-          cts:search( collection('customers'),
-              string($source-node/path/to/lookup/key))/id
-    or populate the instance with constants.
-    =>map:with('constantValue', 10)
-    The output of this function should structurally match the output of
-    es:model-get-test-instances($model)
-    :)
-
-    $instance
+    if (empty($source-node/*)) 
```

Comment:
Well, I agree this looks rather inelegant, and I tripped up on it before.

init-instance returns a structure like this if the source is a reference  (a source is assumed to be a reference if there are no children of the type node... ie `<friendOf><Person>112</Person></friendOf>` would be extracted returned from es:init-instance() as 

`{"$type":"Person":"$ref":112}`

So if the return value from init-instance has this "$ref" key, I don't want to process all of the setters.